internal/util/builder/kongstateservicebackend.go

```diff
@@ -1,48 +1,84 @@
 package builder
 
 import (
+	"fmt"
+
 	"github.com/samber/lo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	"github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/kongstate"
 )
 
 // KongstateServiceBackendBuilder is a builder for KongstateServiceBackend.
 // Primarily used for testing.
 type KongstateServiceBackendBuilder struct {
-	kongstateServiceBackend kongstate.ServiceBackend
+	name      string
+	namespace string
+	weight    *int32
+	portDef   kongstate.PortDef
+	t         kongstate.ServiceBackendType
 }
 
 func NewKongstateServiceBackend(name string) *KongstateServiceBackendBuilder {
 	return &KongstateServiceBackendBuilder{
-		kongstateServiceBackend: kongstate.ServiceBackend{
-			Name: name,
-		},
+		name: name,
 	}
 }
 
 func (b *KongstateServiceBackendBuilder) WithNamespace(namespace string) *KongstateServiceBackendBuilder {
-	b.kongstateServiceBackend.Namespace = namespace
+	b.namespace = namespace
 	return b
 }
 
 func (b *KongstateServiceBackendBuilder) WithWeight(weight int) *KongstateServiceBackendBuilder {
-	b.kongstateServiceBackend.Weight = lo.ToPtr(int32(weight))
+	b.weight = lo.ToPtr(int32(weight))
 	return b
 }
 
 func (b *KongstateServiceBackendBuilder) WithPortNumber(port int) *KongstateServiceBackendBuilder {
-	b.kongstateServiceBackend.PortDef = kongstate.PortDef{
+	b.portDef = kongstate.PortDef{
 		Number: int32(port),
 		Mode:   kongstate.PortModeByNumber,
 	}
 	return b
 }
 
+func (b *KongstateServiceBackendBuilder) WithPortName(port string) *KongstateServiceBackendBuilder {
+	b.portDef = kongstate.PortDef{
+		Name: port,
+		Mode: kongstate.PortModeByName,
+	}
+	return b
+}
+
 func (b *KongstateServiceBackendBuilder) WithType(t kongstate.ServiceBackendType) *KongstateServiceBackendBuilder {
-	b.kongstateServiceBackend.Type = t
+	b.t = t
 	return b
 }
 
 func (b *KongstateServiceBackendBuilder) Build() kongstate.ServiceBackend {
-	return b.kongstateServiceBackend
+	// Default to Kubernetes Service backend type if not specified.
+	if b.t == "" {
+		b.t = kongstate.ServiceBackendTypeKubernetesService
+	}
+	// Default to default namespace if not specified.
+	if b.namespace == "" {
+		b.namespace = metav1.NamespaceDefault
+	}
+
+	s, err := kongstate.NewServiceBackend(
+		b.t,
+		b.namespace,
+		b.name,
+		b.portDef,
+	)
+	if err != nil {
+		// This should never happen. If it does, it's a bug that will be discovered in tests as the builder
+		// is used in tests only.
+		panic(fmt.Errorf("failed to build service backend: %w", err))
```

Comment:
I think it's good approach, but what I'd like suggest to rename `Build()` to `MustBuild()` it's common idiom in Go to indicate that in case of wrong input, function will panic

Reply:
Yup, good point. Should discourage from using it outside of tests. 👍